[FIX] Allow to deactivate and restore quick creation without restarting

StefanRijnhart opened the pull request:
The implementation marks seen model objects with a special attribute to prevent unnecessary processing.  If a model was modified, _register_hook was called but the attribute prevented quick creation to be disabled. Also, re-enabling quick creation at run time was not supported.

StefanRijnhart committed:
[ADD] README.rst

hbrunn commented on base_optional_quick_create/model.py:
```diff
@@ -38,16 +39,42 @@ def wrapper(cr, uid, name, context=None):
+                    logger.debug(
+                        'Replacing original name_create method on %s',
+                        model.model)
+                    model_obj.name_create_orig = model_obj.name_create
```
how about using `_patch_method`? https://github.com/OCA/OCB/blob/9.0/openerp/models.py#L5268

hbrunn commented on base_optional_quick_create/model.py:
```diff
@@ -38,16 +39,42 @@ def wrapper(cr, uid, name, context=None):
+                    model_obj.name_create = model_obj.name_create_orig
+                    delattr(model_obj, 'name_create_orig')
+            model_obj.check_quick_create = True
         return True
```
return super's result

hbrunn commented on base_optional_quick_create/model.py:
```diff
@@ -38,16 +39,42 @@ def wrapper(cr, uid, name, context=None):
+                    logger.debug(
+                        'Restoring original name_create method on %s',
+                        model.model)
+                    model_obj.name_create = model_obj.name_create_orig
```
if you take my comment above, you can use `_revert_method` here: https://github.com/OCA/OCB/blob/9.0/openerp/models.py#L5299

hbrunn commented on base_optional_quick_create/model.py:
```diff
@@ -38,16 +39,42 @@ def wrapper(cr, uid, name, context=None):
         return wrapper
 
     def _register_hook(self, cr, ids=None):
```
I think it's dangerous to change `_register_hook`'s signature. Better refactor this into its own method and just call it here

StefanRijnhart replied:
That trick was taken from addons/base_action_rule/base_action_rule.py

StefanRijnhart replied:
But actually I agree with you.

eLBati replied:
@hbrunn @StefanRijnhart what problem do you see with adding `ids=None`?

StefanRijnhart replied:
@eLBati let's say another module overrides _register_hook for the same class. Its signature does not contain the ids argument, because it is identical to the signature in core Odoo. Your module then calls _register_hook with the new ids argument. If the other module happens to come first, this will cause a crash.

eLBati replied:
@StefanRijnhart @hbrunn I would not know how to modify `_register_hook` keeping its behaviour.
What do you think about describing this incompatibility in README?

StefanRijnhart replied:
Just cooked up this (untested): https://github.com/OCA/server-tools/pull/340/files

pedrobaeza commented:
Why don't expand https://github.com/OCA/web/tree/8.0/web_m2x_options? In my opinion, this is more a web client issue than a base/server one. You can take the idea of creating ir.parameter per model like in this module: https://github.com/OCA/social/blame/9.0/mail_restrict_follower_selection/README.rst#L13

StefanRijnhart commented:
The advantage of this method is that you can set the option per model
instead of per form. It also allows the administrator user to configure
each model, instead of having to add options to all the fields in all
the views (which is how I remember that module works).

pedrobaeza commented:
Yeah, I know, but I'm talking about implementing this granularity in that module, and make it at web client level, where it should be (as is this web client who includes the quick create option, not the ORM). Obviously, this requires to be done in JS.

pedrobaeza commented:
I didn't see that this module already exists, but for me it is a mistake to be implemented this way.

StefanRijnhart commented:
@pedrobaeza maybe you are right but can we keep this module in its current form in the 8.0 branch for the sake of continuity?

pedrobaeza commented:
OK, but please put a warning about this in the Known issues for the v9 migration.

eLBati commented:
@StefanRijnhart as https://github.com/OCA/server-tools/pull/352 is merged, what is the destiny of this?

StefanRijnhart committed:
[IMP] Reallow quick create on a model without restarting the application
[RFR] Don't keep our own check attribute, use the patch mechanism's 'origin' attribute instead
[RFR] Only pass models to the patch method that need patching

StefanRijnhart commented:
@elBati I have rebased

sbidoul commented:
:+1: 

Can we merge this in 8.0 at least?

eLBati commented:
:+1: thanks
